MeshCoord class [AVD-1591]

pp-mo opened the pull request:
## 🚀 Pull Request

### Description
Adding MeshCoord.

Notes: 
  1. the characteristic points and bounds indexing calculations are **_not_** yet implemented here.
I have other code prototyped for that, but this PR is simpler if I leave it out for now.
  2. the implementation isn't quite minimal, but I have chosen to control copying and slicing to make it a 'good citizen' within a cube (i.e. not blocking cube copy or slicing).
  3. there is as yet no sensible `__str__` or `__repr__`.  We have another task pending to address that.
  
---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

pp-mo commented:
I think this is all that is needed, but you may well decide there is something missing @trexfeathers.
The testing is using specific real things rather than mock-ist, so you might well uncover some possibility I have missed due to that.

I am a little surprised there is not more to this, but that is probably because we have decided now to defer implementing the "bounds but no points" ability.  Which I thought might be tricky, and now think even more so!

@stephenworsley it will be good if you can cast an eye over this, but I think @trexfeathers is steering at this point.

pp-mo commented:
Update: I'm now working to reduce the `__init__` args to just mesh/location/axis, and thus enforce the intended content wrt the provided args.

I'll "WIP" / draft this : please ***don't merge*** till those changes are up

trexfeathers commented on lib/iris/experimental/ugrid.py:
```diff
@@ -2115,6 +2116,252 @@ def face_node(self):
+
+    """
+
+    # Note: effectively this list of valid mesh locations is determined by
+    # the Mesh api, but it doesn't expose those choices in a generic form, so
+    # we will just define our own here.
```
Could add a `# TODO:` comment? I've been doing that with my data model work so far.

pp-mo replied:
ok

pp-mo commented:
Failing docs tests -- probably needs another rebase ?

Meantimes, since discussing with @trexfeathers today, I am now intending to fix the copy/init to support the cube slicing needs in a nicer way (i.e. without bodging the allowed copy  args as at present).
Hence this is still WIP

trexfeathers reviewed (changes requested):
I've been asked to stop reviewing this and work on something else. So here's what I've got so far. Mostly optional minor improvements, but I definitely think you should remove those `setter` methods (see individual comments). Thanks 🙂

pp-mo commented on lib/iris/experimental/ugrid.py:
```diff
@@ -2115,6 +2116,242 @@ def face_node(self):
+    """
+    Geographic coordinate values of data on an unstructured mesh.
+
+    A MeshCoord contains a `~iris.experiomental.ugrid.Mesh`.
```
ok

pp-mo commented on lib/iris/experimental/ugrid.py:
```diff
@@ -2115,6 +2116,242 @@ def face_node(self):
+
+    A MeshCoord also specifies its 'axis' : 'x' or 'y'.  Its values are then,
+    accordingly, longitudes or latitudes.  The values are taken from the
+    appropriate element coordinates and connectivities in the Mesh, determined
```
ok, I think we can simply remove the word 'element'

pp-mo commented on lib/iris/experimental/ugrid.py:
```diff
@@ -2115,6 +2116,242 @@ def face_node(self):
+
+    .. note::
+        As described above, it is possible for a MeshCoord to have bounds but
+        no points.  This is not possible for a regular AuxCoord or DimCoord.
```
ok

pp-mo commented on lib/iris/experimental/ugrid.py:
```diff
@@ -2115,6 +2116,242 @@ def face_node(self):
+        # created by methods of the Mesh itself, which will ensure correctness.
+        if not isinstance(mesh, Mesh):
+            msg = (
+                "'mesh' must be an iris.experimental.ugrid.Mesh, "
```
ok!

pp-mo commented on lib/iris/experimental/ugrid.py:
```diff
@@ -2115,6 +2116,242 @@ def face_node(self):
+    def mesh(self):
+        return self._mesh
+
+    @mesh.setter
+    def mesh(self, value):
+        msg = "Cannot set the 'mesh' of a MeshCoord."
+        raise ValueError(msg)
```
:+1: good spot, will do !

pp-mo commented on lib/iris/experimental/ugrid.py:
```diff
@@ -2115,6 +2116,242 @@ def face_node(self):
+        """The coordinate-system of a MeshCoord is always 'None'."""
+        return None
+
+    @coord_system.setter
+    def coord_system(self, value):
+        if value is not None:
+            msg = "Cannot set the coordinate-system of a MeshCoord."
+            raise ValueError(msg)
```
This one + 'climatalogical' are different :  They must be writable by the parent constructor.

pp-mo replied:
I think I have addressed this now as far as I can.
I have removed setters the mesh/location/axis properties.
But I need them for coord_system / climatological, because the parent constructor writes them.
(if I remove them, I get errors).

pp-mo commented on lib/iris/experimental/ugrid.py:
```diff
@@ -2115,6 +2116,242 @@ def face_node(self):
+        """The 'climatological' of a MeshCoord is always 'False'."""
+        return False
+
+    @climatological.setter
+    def climatological(self, value):
+        if value:
+            msg = "Cannot set 'climatological' on a MeshCoord."
+            raise ValueError(msg)
```
'coord_system' + 'climatalogical' are different :  
They do need a setter, as they are written by the parent constructor code.

pp-mo commented on lib/iris/experimental/ugrid.py:
```diff
@@ -2115,6 +2116,242 @@ def face_node(self):
+    def location(self):
+        return self._metadata_manager.location
+
+    @location.setter
+    def location(self, value):
+        msg = "Cannot set the 'location' of a MeshCoord."
+        raise ValueError(msg)
```
ok

pp-mo commented on lib/iris/experimental/ugrid.py:
```diff
@@ -2115,6 +2116,242 @@ def face_node(self):
+    def axis(self):
+        return self._metadata_manager.axis
+
+    @axis.setter
+    def axis(self, value):
+        msg = "Cannot set the 'axis' of a MeshCoord."
+        raise ValueError(msg)
```
ok

pp-mo commented:
Latest:
 * rebased to fix docs test failures (I hope)
 * addressed all outstanding comments (I think)
 * revised the `copy` and `__getitem__` overrides to support minimum needs, as promised above
 * added testing for the result points+bounds arrays  (preliminary, pending "dynamic" versions of the calcs)
 
Please review @bjlittle !

bjlittle commented on lib/iris/tests/unit/experimental/ugrid/test_MeshCoord.py:
```diff
@@ -0,0 +1,394 @@
+        # The property exists, =None, can set to None, can not set otherwise.
+        self.assertTrue(hasattr(self.meshcoord, "coord_system"))
+        self.assertIsNone(self.meshcoord.coord_system)
+        self.meshcoord.coord_system = None
```
@pp-mo This is slightly confusing... you shouldn't be able to set `coord_system` to anything, it should be completely read-only.

pp-mo replied:
Please see notes  above.
I provide a setter, and allow it to write ONLY None.
Because the parent constructor requires to write this, unless we rewrite it somehow.  I really don't **_want_** to change the method by which this is set, since in all other cases it is part of the metadata so the mechanism is provided 'automatically'.

bjlittle commented on lib/iris/tests/unit/experimental/ugrid/test_MeshCoord.py:
```diff
@@ -0,0 +1,394 @@
+        # The property exists, =False, can set to False, can not set otherwise.
+        self.assertTrue(hasattr(self.meshcoord, "climatological"))
+        self.assertFalse(self.meshcoord.climatological)
+        self.meshcoord.climatological = False
```
As above

bjlittle commented on lib/iris/tests/unit/experimental/ugrid/test_MeshCoord.py:
```diff
@@ -0,0 +1,394 @@
+        meshcoord = _create_test_meshcoord()
+        data_shape = (2,) + _TEST_POINTS.shape
+        cube = Cube(np.zeros(data_shape))
+        cube.add_aux_coord(meshcoord, 1)
```
I guess if this goes wrong, it goes wrong here in the `setup` :+1:

pp-mo replied:
Yes, that's just why `test_basic` appears to be a bit "thin" !

bjlittle commented on lib/iris/tests/unit/experimental/ugrid/test_MeshCoord.py:
```diff
@@ -0,0 +1,394 @@
+        cube = self.cube
+        self.assertIs(cube.coord(standard_name="longitude"), meshcoord)
+        self.assertIs(cube.coord(long_name="long-name"), meshcoord)
+        # self.assertIs(cube.coord(var_name="var"), meshcoord)
```
@pp-mo Nuke this? Or was it a reminder to do some additional testing specifically for `var_name`?

pp-mo replied:
No, it's dead code left over from a previous approach. 
(!)

bjlittle reviewed (changes requested):
@pp-mo Lovely stuff :+1: 

This is the super interesting part (for me) now... the interaction of `MeshCoord`, `Mesh` and `Cube`

Okay, so I've raised a few review questions and issues, but nothing controversial . I'm sure you can easily address them.

Otherwise, this is looking very promising :+1:

pp-mo commented on lib/iris/experimental/ugrid.py:
```diff
@@ -51,6 +51,7 @@
     "MeshNodeCoords",
     "MeshMetadata",
     "MeshCoordMetadata",
+    "MeshCoord",
```
ok

pp-mo commented on lib/iris/experimental/ugrid.py:
```diff
@@ -2680,6 +2681,257 @@ def face_node(self):
+    """
+    Geographic coordinate values of data on an unstructured mesh.
+
+    A MeshCoord references a `~iris.experiomental.ugrid.Mesh`.
```
:+1:

pp-mo commented on lib/iris/experimental/ugrid.py:
```diff
@@ -2680,6 +2681,257 @@ def face_node(self):
+    A MeshCoord references a `~iris.experiomental.ugrid.Mesh`.
+    When contained in a `~iris.cube.Cube` it connects the cube to the Mesh,
+    and records (a) which cube dimension is mapped to the mesh, and (b) on
+    which mesh location (e.g. 'face' or 'node') the cube data exists.
```
:+1:

pp-mo commented on lib/iris/experimental/ugrid.py:
```diff
@@ -2680,6 +2681,257 @@ def face_node(self):
+    # Note: the valid mesh locations are effectively part of the Mesh api, but
+    # it doesn't provide them in a simple form, so we define our own here.
+    # TODO: make this a public item of the Mesh class ?
+    VALID_MESH_LOCATIONS = ("face", "edge", "node")
```
:+1:
will mean yet another a rebase :disappointed:

pp-mo commented on lib/iris/tests/unit/experimental/ugrid/test_MeshCoord.py:
```diff
@@ -0,0 +1,394 @@
+            _create_test_meshcoord(mesh=mock.sentinel.odd)
+
+    def test_valid_locations(self):
+        for loc in ("face", "edge", "node"):
```
:+1:

pp-mo commented:
Thanks @bjlittle 
Rebased to use your recent addition of `Mesh.LOCATIONS`.
Hope I've addressed everything raised so far.
See if you still don't like the setters for coord_system + climatological .  If you really dislike this, I guess we can do something different with these somehow, but it's a bit "tail wagging dog" to modify Coord for the benefit of the subclass !

trexfeathers commented on lib/iris/experimental/ugrid.py:
```diff
@@ -2680,6 +2681,257 @@ def face_node(self):
+        # Get the 'coord identity' metadata from the relevant node-coordinate.
+        # N.B. mesh.coord returns a dict
+        node_coords = self.mesh.coord(include_nodes=True, axis=self.axis)
+        (node_coord,) = list(node_coords.values())
```
What about:

```python
node_coord = self.mesh.node_coords._asdict[f"node_{self.axis}"]
```

pp-mo commented:
@bjlittle apologies, forgot to push.
Please re-review :grin:

bjlittle commented on lib/iris/experimental/ugrid.py:
```diff
@@ -2680,6 +2681,257 @@ def face_node(self):
+                f"{Mesh.__module__}.{Mesh.__name__}, "
+                f"got {mesh}."
+            )
+            raise ValueError(msg)
```
@pp-mo You've not addressed this yet

pp-mo replied:
:+1: nice spot!

bjlittle commented on lib/iris/experimental/ugrid.py:
```diff
@@ -2680,6 +2681,257 @@ def face_node(self):
+
+        points, bounds = self._construct_access_arrays()
+        if points is None:
+            # We intend to support this in future, but it will require extra
```
@pp-mo You've not addressed this yet

pp-mo replied:
:+1:

bjlittle commented on lib/iris/experimental/ugrid.py:
```diff
@@ -2680,6 +2681,257 @@ def face_node(self):
+        # warning may return NotImplemented or a numpy boolean.
+        if eq is not NotImplemented and eq:
+            # Ensures that 'other' is also a MeshCoord
+            eq = self.mesh == other.mesh
```
@pp-mo You've not addressed this yet

pp-mo replied:
I hadn't spotted this.
I think we need to do more : "Ideally", if we accept that the meshes are equivalent, then we know that the points, bounds and basic coord properties (names, units, attributes) *must* all be equal
Except, actually, we **_don't_**, since the Mesh might have changed between creating the two MeshCoords :cry: 

So, I want to say we should *never* check anything but the Mesh/location/axis, certainly not points, bounds, attributes etc...
But owing to the limitations on the 'dynamic'-ness, this isn't actually quite solid.

So, what do you think we **_should_** we do with this?

pp-mo replied:
Stop press: 
I'm now proposing we compare mesh/location/axis.
That makes logical sense, though it is not actually quite correct in view of the possibility for the Mesh to change.

bjlittle commented on lib/iris/tests/unit/experimental/ugrid/test_MeshCoord.py:
```diff
@@ -0,0 +1,394 @@
+import numpy as np
+import unittest.mock as mock
+
+from iris.cube import Cube
```
@pp-mo You've not addressed this yet

pp-mo replied:
:+1:

bjlittle commented on lib/iris/tests/unit/experimental/ugrid/test_MeshCoord.py:
```diff
@@ -0,0 +1,394 @@
+_TEST_N_NODES = 15
+_TEST_N_FACES = 3
+_TEST_N_EDGES = 5
+_TEST_N_BOUNDS = 4
```
@pp-mo You've not addressed this yet

bjlittle commented on lib/iris/tests/unit/experimental/ugrid/test_MeshCoord.py:
```diff
@@ -0,0 +1,394 @@
+    def test_derived_properties(self):
+        # Check the derived properties of the meshcoord against the correct
+        # underlying mesh coordinate.
+        for axis in ("x", "y"):
```
@pp-mo You've not addressed this yet

pp-mo replied:
:+1:

pp-mo commented:
Further changes as requested (?mostly).
Please re-view @bjlittle 

I'm still slightly uncomfortable about the shortcut on `__eq__` comparison, but we clearly **_don't_** want it comparing points+bounds also.

pp-mo commented:
Fixed the equality thing.
As discussed offline.
Ok @bjlittle !

bjlittle commented on lib/iris/experimental/ugrid.py:
```diff
@@ -2884,20 +2884,13 @@ def __deepcopy__(self, memo):
-                    and self.axis == other.axis
-                )
+                # Compare rest of metadata, but not points/bounds.
+                eq = eq and self.metadata == other.metadata
```
@pp-mo `eq` doesn't need to be factored in at this point, thus this simple reduces to:
```python
eq = self.metadata == other.metadata
```
:+1:

pp-mo replied:
:+1:

bjlittle commented on lib/iris/tests/unit/experimental/ugrid/test_MeshCoord.py:
```diff
@@ -178,6 +179,35 @@ def test_set_climatological(self):
+                meshcoord2.units = "Pa"
+            elif prop == "attributes":
+                meshcoord2.attributes["thing"] = 77
+        self.assertNotEqual(meshcoord2, meshcoord)
```
Nice one :+1:
